pkg/redfish/redfish-foxconn.go

```diff
@@ -0,0 +1,323 @@
+	jsonData, err := s.epRF.GETRelative(path)
+	if err != nil || jsonData == nil {
+		s.LastStatus = HTTPsGetFailed
+		return
+	}
+	s.LastStatus = HTTPsGetOk
+
+	var n InsydeOemNcsiCollection
+	if err := json.Unmarshal(jsonData, &n); err != nil {
+		errlog.Printf("Failed to decode %s: %s\n", url, err)
+		s.LastStatus = EPResponseFailedDecode
+	}
+
+	if n.MembersOCount < 1 {
+		errlog.Printf("No Ncsi members detected")
+		return
+	}
+
+	//////////////////////////////////////////////////////
+	// Parse each /redfish/v1/Systems/system/Oem/Insyde/Ncsi/# entry
+
+	for _, ncsiMember := range n.Members {
+		path := ncsiMember.Oid
+
+		url := s.epRF.FQDN + path
+		jsonData, err = s.epRF.GETRelative(path)
+		if err != nil || jsonData == nil {
+			s.LastStatus = HTTPsGetFailed
+			return
+		}
+		s.LastStatus = HTTPsGetOk
+
+		var nm InsydeOemNcsiMember
+		if err := json.Unmarshal(jsonData, &nm); err != nil {
+			errlog.Printf("Failed to decode %s: %s\n", url, err)
+			s.LastStatus = EPResponseFailedDecode
+		}
+
+		//////////////////////////////////////////////////////
+		// Parse each /redfish/v1/Systems/system/Oem/Insyde/Ncsi/# package member element
+		// Have only ever seen one package but we should iterate anyway in case that ever changes
+
+		for _, nmPkg := range nm.Package {
+
+			//////////////////////////////////////////////////////
+			// Parse /redfish/v1/Systems/system/Oem/Insyde/Ncsi/#/Package/#
+
+			path = nmPkg.Oid
+
+			url = s.epRF.FQDN + path
+			jsonData, err = s.epRF.GETRelative(path)
+			if err != nil || jsonData == nil {
+				s.LastStatus = HTTPsGetFailed
+				return
+			}
+			s.LastStatus = HTTPsGetOk
+
+			var p InsydeOemPackage
+			if err := json.Unmarshal(jsonData, &p); err != nil {
+				errlog.Printf("Failed to decode %s: %s\n", url, err)
+				s.LastStatus = EPResponseFailedDecode
+			}
+
+			//////////////////////////////////////////////////////
+			// Parse each /redfish/v1/Systems/system/Oem/Insyde/Ncsi/#/Package/#.PackageInfo[]
+
+			for j, pi := range p.PackageInfo {
+				// Only process if there is a MAC address
+				if pi.MACAddress != "" {
+					s.ENetInterfaces.Num++
+
+					ei := NewEpEthInterface(s.epRF, ncsiMember.Oid, s.RedfishSubtype, nmPkg, s.ENetInterfaces.Num)
+
+					ei.EtherIfaceRF.Oid = path
+					ei.EtherIfaceRF.MACAddress = pi.MACAddress
+					ei.EtherIfaceRF.Description = FOXCONN_ETH_INT_DESCRIPTION
+
+					// ID = "ncsi-" + ncsi number + "-" + package number + "-" + channel index
+					ei.EtherIfaceRF.Id = "ncsi" + nm.Id + "-p" + p.Id + "-c" + fmt.Sprint(j)
+
+					// According to Foxconn, this is the only unique identifier for the primary ethernet
+					if nm.VersionId.PCIDID == FOXCONN_PRIMARY_ETH_INT_PCIDID {
+						// We append a unique string to the end of the ID so that we can identify the
+						// primary ethernet interface later.
+						ei.EtherIfaceRF.Id += FOXCONN_PRIMARY_ETH_INT_SUFFIX
+					} else if strings.TrimSpace(nm.VersionId.FirmwareName) == FOXCONN_PRIMARY_ETH_INT_FW_NAME {
```

Comment:
The whole point of this block is to leave behind a breadcrumb if we detect a different PCIDID for the primary ethernet.  There are two issues with this though...

Issue one assumes that the FirmwareName of "X550 FW Ver" is unique to the primary ethernet and will never change.  In one of the PRDIS team's Paradise tester nodes I saw, what I assume were Mellonox cards, have the "mlx0.1" FirmwareName.

Issue two is if Foxconn ever changes the FirmwareName.  The actual value in the systems we have today read ""X550 FW Ver " with a space appended to the end of the string after "Ver".  I could easily see that being a typo that they might fix in the future so used strings.TrimSpace to remove it before comparing.

If this ever fails here we'll simply leave behind an appropriate breadcrumb.